cmd/argocd/commands/app.go

```diff
@@ -1635,35 +1713,42 @@ func waitOnApplicationStatus(acdClient apiclient.Client, appName string, timeout
 		var selectedResourcesAreReady bool
 
 		// If selected resources are included, wait only on those resources, otherwise wait on the application as a whole.
-		if len(selectedResources) > 0 {
+		if len(opts.selectedResources) > 0 {
 			selectedResourcesAreReady = true
-			for _, state := range getResourceStates(app, selectedResources) {
-				resourceIsReady := checkResourceStatus(watchSync, watchHealth, watchOperation, watchSuspended, state.Health, state.Status, appEvent.Application.Operation)
+			for _, state := range getResourceStates(app, opts.selectedResources) {
+				resourceIsReady := checkResourceStatus(opts.watch, state.Health, state.Status, appEvent.Application.Operation)
 				if !resourceIsReady {
 					selectedResourcesAreReady = false
 					break
 				}
 			}
 		} else {
 			// Wait on the application as a whole
-			selectedResourcesAreReady = checkResourceStatus(watchSync, watchHealth, watchOperation, watchSuspended, app.Status.Health.Status, string(app.Status.Sync.Status), appEvent.Application.Operation)
+			selectedResourcesAreReady = checkResourceStatus(opts.watch, app.Status.Health.Status, string(app.Status.Sync.Status), appEvent.Application.Operation)
 		}
 
-		if selectedResourcesAreReady {
+		if selectedResourcesAreReady && !opts.watch.delete {
 			printFinalStatus(app)
 			return app, nil
 		}
 
-		newStates := groupResourceStates(app, selectedResources)
+		newStates := groupResourceStates(app, opts.selectedResources)
 		for _, newState := range newStates {
 			var doPrint bool
 			stateKey := newState.Key()
 			if prevState, found := prevStates[stateKey]; found {
-				if watchHealth && prevState.Health != argoappv1.HealthStatusUnknown && prevState.Health != argoappv1.HealthStatusDegraded && newState.Health == argoappv1.HealthStatusDegraded {
+				if opts.watch.health && prevState.Health != argoappv1.HealthStatusUnknown && prevState.Health != argoappv1.HealthStatusDegraded && newState.Health == argoappv1.HealthStatusDegraded {
 					printFinalStatus(app)
-					return nil, fmt.Errorf("Application '%s' health state has transitioned from %s to %s", appName, prevState.Health, newState.Health)
+					return nil, fmt.Errorf("Application '%s' health state has transitioned from %s to %s", opts.appName, prevState.Health, newState.Health)
+				}
+				if opts.watch.delete {
+					if newState.Health == "Missing" {
+						newState.Message = strings.Replace(newState.Message, "created", "deleted", 1)
+						doPrint = prevState.Merge(newState)
+					}
```

Comment:
Your strategy so far seems to spawn a new goroutine to poll and see if the application has been deleted. I don't think that is necessary since the application as a whole will not be deleted until all of its resources are deleted first. Furthermore, we already have a channel open into the application events. No need to open a new one.

My suggested strategy is this:

1. First wait until all of the `newState.Health` are Missing (deleted). If at least one of them is not, keep waiting until they all are.
2. Once all of the `newState.Health` are missing, check if the application as a whole is missing.

If both 1 and 2 are true, then you can return. These should be all the changes that need to be made.

Here is a little sketch:
```golang
newStates := groupResourceStates(app, opts.selectedResources)
// This flag checks if an application is deleted. If `opts.watch.delete` is set to false, then we don't care if it is or not.
applicationIsDeleted := opts.watch.delete
for _, newState := range newStates {
	var doPrint bool
	if opts.watch.delete {
                // Even if a single resource is not deleted, then the application as a whole will not be
		if newState.Health != argoappv1.HealthStatusMissing {
			applicationIsDeleted = false
		}
	}
        ...
}
// If all resources are deleted, _then_ we can check if the application as a whole is also deleted.
if applicationIsDeleted {
	_, err := appClient.Get(context.Background(), &applicationpkg.ApplicationQuery{Name: &opts.appName})
	if err != nil {
		if strings.HasSuffix(err.Error(), "not found") {
			return nil, nil
		}
	}
}
```

Reply:
Also you should use `argoappv1.HealthStatusMissing` instead of the literal `"Missing"`.

Reply:
@adamjohnson01 I edited this comment, please take a look!

Reply:
I was doing this but I thought that if there are a lot of resources and events there could be a lot of Get calls and thought it may be better to check outside of the events. Do you think it is ok to do the check for every event?

Reply:
If we are going to go with a polling approach I think polling based on events is preferable to polling based on time intervals since it is constrained to the application advancing towards a deleted state. If we polled based on time intervals and the deletion operation stalled or took an exceptionally long time, the default behavior would be to keep polling endlessly (since there is no timeout by default). On events this wouldn't happen.

Ideally there would be an `ApplicationWatchEvent` that signifies the application deleted successfully. In that case we could simply wait for it to be passed on the channel.

Reply:
How about we check if the event channel has been closed and then check if the application has been deleted? Instead of checking after every event?

Reply:
I don't think the channel closes upon application deletion currently.

Reply:
OK, I have put the check back inside the events loop.